### PR TITLE
updating esm2 native recipe

### PR DESCRIPTION
**Pull request by pstjohn**
* Adds separate `train_ddp.py`, `train_fsdp2.py`, and `train_nvfsdp.py` entrypoints
* Adds comparison against FA-2 based HF transformers model

## Summary by CodeRabbit

* **New Features**
  * Distributed ESM‑2 training entrypoints (DDP, MFSDP/FSDP2) and a shared linear warmup/decay LR scheduler.

* **Configuration**
  * Switched defaults to nvFSDP-style sharding; updated model identifiers and training hyperparameters (train steps, warmup, optimizer LR).

* **Documentation**
  * Added new ESM‑2 training README; removed an outdated README.

* **Build/Chores**
  * Install Transformers from the Git repo; Docker builds can use netrc credentials for installs.

* **Tests**
  * Added extensive single‑ and multi‑GPU training tests; removed obsolete tests.

### CODE REVIEW

File: recipes/esm2_native_te_mfsdp/train_fsdp2.py [hunk 1]
```diff
@@ -106,27 +106,23 @@ def main(args: DictConfig):
...
-    #     f"nvidia/{args.model_name}", trust_remote_code=True, torch_dtype=torch.bfloat16
-    # )
-    # config.max_seq_length = args.max_seq_length
-    # config.micro_batch_size = args.micro_batch_size
-
-    # with torch.device("meta"):
-    #     model = AutoModelForMaskedLM.from_config(config, trust_remote_code=True)
-
-    for layer in model.esm.encoder.layer:
+    if "facebook" in args.model_name:
```
Comment (jomitchellnv) [inline recipes/esm2_native_te_mfsdp/train_fsdp2.py:86]: do we need to update the other ESM2 recipes as well?

File: recipes/esm2_native_te_mfsdp/train_fsdp2.py [hunk 2]
```diff
@@ -139,10 +135,11 @@ def main(args: DictConfig):
     optimizer = AdamW(model.parameters(), **args.adamw_kwargs)
     scheduler = get_linear_schedule_with_warmup(optimizer, **args.lr_scheduler_kwargs)
 
-    model.to_empty(device=f"cuda:{dist_config.local_rank}")
-    for module in model.modules():
-        if hasattr(module, "reset_parameters"):
-            module.reset_parameters()
+    model.to(device=device)
+    # model.to_empty(device=f"cuda:{dist_config.local_rank}")
```
Comment (jomitchellnv) [inline recipes/esm2_native_te_mfsdp/train_fsdp2.py]: comment

**Review by jomitchellnv — APPROVED** [review]
LGTM

File: recipes/esm2_native_te_mfsdp/train_ddp.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    optimizer = AdamW(model.parameters(), **args.adamw_kwargs)
+    scheduler = get_linear_schedule_with_warmup(optimizer, **args.lr_scheduler_kwargs)
+
+    model = model.to(device=device)
+    model = torch.nn.parallel.DistributedDataParallel(
+        model,
+        device_ids=[dist_config.local_rank],
+        output_device=dist_config.local_rank,
+        device_mesh=device_mesh["fsdp"],
+    )
```
[the remark this answers is not in the document]
Comment (pstjohn) [reply recipes/esm2_native_te_mfsdp/train_ddp.py]: that's not true

File: recipes/esm2_native_te_mfsdp/train_fsdp2.py (hunk 1 above)
Comment (pstjohn) [reply recipes/esm2_native_te_mfsdp/train_fsdp2.py:86]: using this as an easy flag to swap TE / non-TE training is nice; sure

File: recipes/esm2_native_te_mfsdp/hydra_config/L0_sanity.yaml [hunk 4]
```diff
@@ -13,5 +13,7 @@ wandb_init_args:
 
 # Learning rate scheduler config
 lr_scheduler_kwargs:
-  num_warmup_steps: 2
-  num_training_steps: 8
+  num_warmup_steps: 0
```
[the remark this answers is not in the document]
Comment (pstjohn) [reply recipes/esm2_native_te_mfsdp/hydra_config/L0_sanity.yaml:16]: it comes from the defaults.yaml

File: recipes/esm2_native_te_mfsdp/train_fsdp2.py (hunk 2 above)
Comment (pstjohn) [reply recipes/esm2_native_te_mfsdp/train_fsdp2.py]: removed; we now use the to_empty path

**Review by jstjohn — APPROVED** [review]

**Comment by trvachov** [comment]
Closing as stale — this concerned ESM2 batch_size validation. The `bionemo-esm2` sub-package has been removed and ESM2 was rewritten in `bionemo-recipes/models/` with a new architecture.
